File: Leetcode/tree/437_PathSumIII/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: int
        """
        if root is None:
            return 0
        return  self.count(root, sum) + self.pathSum(root.left, sum) + self.pathSum(root.right, sum)

    def count(self, root, sum):
        if root is None:
            return 0
        c = 1 if root.val==sum else 0
        return c + self.count(root.left, sum-root.val) + self.count(root.right, sum-root.val)

File: Leetcode/tree/437_PathSumIII/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    int count(TreeNode* root, int sum){
        if(root==nullptr) return 0;
        return ((root->val==sum)?1:0)+count(root->left, sum-root->val)+count(root->right, sum-root->val);
    }
public:
    int pathSum(TreeNode* root, int sum) {
        if(root==nullptr) return 0;
        return count(root, sum)+pathSum(root->left, sum)+pathSum(root->right, sum);
    }
};
